pkg/controller/clusterpool/clusterpool_controller_test.go

```diff
@@ -1251,6 +1251,127 @@ func TestReconcileClusterPool(t *testing.T) {
+				testclaim.FullBuilder(testNamespace, "test-claim", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+			},
+			expectedObservedSize:   4,
+			expectedObservedReady:  4,
+			expectedTotalClusters:  5,
+			expectedRunning:        3,
+			expectedAssignedCDs:    1,
+			expectedAssignedClaims: 1,
+		},
+		{
+			name: "pool size > number of claims > runningCount",
+			existing: []runtime.Object{
+				initializedPoolBuilder.Build(
+					testcp.WithSize(4),
+					testcp.WithRunningCount(2),
+				),
+				unclaimedCDBuilder("c1").Build(testcd.WithPowerState(hivev1.RunningClusterPowerState), testcd.Installed()),
+				unclaimedCDBuilder("c2").Build(testcd.WithPowerState(hivev1.RunningClusterPowerState), testcd.Installed()),
+				unclaimedCDBuilder("c3").Build(testcd.Installed()),
+				unclaimedCDBuilder("c4").Build(testcd.Installed()),
+				testclaim.FullBuilder(testNamespace, "test-claim1", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+				testclaim.FullBuilder(testNamespace, "test-claim2", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+				testclaim.FullBuilder(testNamespace, "test-claim3", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+			},
+			expectedObservedSize:   4,
+			expectedObservedReady:  4,
+			expectedTotalClusters:  7,
+			expectedRunning:        5,
+			expectedAssignedCDs:    3,
+			expectedAssignedClaims: 3,
+		},
+		{
+			name: "number of claims > pool size > runningCount",
+			existing: []runtime.Object{
+				initializedPoolBuilder.Build(
+					testcp.WithSize(4),
+					testcp.WithRunningCount(2),
+				),
+				unclaimedCDBuilder("c1").Build(testcd.WithPowerState(hivev1.RunningClusterPowerState), testcd.Installed()),
+				unclaimedCDBuilder("c2").Build(testcd.WithPowerState(hivev1.RunningClusterPowerState), testcd.Installed()),
+				unclaimedCDBuilder("c3").Build(testcd.Installed()),
+				unclaimedCDBuilder("c4").Build(testcd.Installed()),
+				testclaim.FullBuilder(testNamespace, "test-claim1", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+				testclaim.FullBuilder(testNamespace, "test-claim2", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+				testclaim.FullBuilder(testNamespace, "test-claim3", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+				testclaim.FullBuilder(testNamespace, "test-claim4", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+				testclaim.FullBuilder(testNamespace, "test-claim5", scheme).Build(testclaim.WithPool(testLeasePoolName)),
+			},
+			expectedObservedSize:  4,
+			expectedObservedReady: 4,
+			expectedTotalClusters: 9,
+			// The four original pool CDs got claimed. Two were already running; we started the other two.
+			// We create five new CDs to satisfy the pool size plus the additional claim.
+			// Of those, we start two for runningCount, and hibernate the rest.
+			// (Intuitively, perhaps the one for the excess claim should start off in running state.
+			// But that's not how the logic works today. Should it?)
```

Comment:
Yeah.

We had to do some [nontrivial work](https://github.com/openshift/hive/pull/1474) for these kinds of scenarios (claims in excess of pool size) to prevent the "overdrawn" clusters from getting deleted to get us back to capacity.

This would behave the same way: those clusters would get `PowerState=Hibernating` until they finished installing. Then -- and this is important -- _the oldest clusters_ (which are not necessarily the ones just installed) would get assigned and set `PowerState=Running`. In fact, those aren't necessarily synchronous steps: if one or more older clusters were already installed/installing, they would be activated and assigned before the surplus ones finished.

It would be nice if, in all cases (regardless of `runningCount`) we could set an extra {number of surplus claims} clusters to running. I'm just worried about the complexity and getting the math right.

So I think I'd like to look into this in a followon. If that's okay with you, I'll open a card.

Reply:
https://issues.redhat.com/browse/HIVE-1651